scan/base.go

```diff
@@ -40,6 +44,378 @@ type base struct {
+	Update  string `json:"update"`
+	Version string `json:"version"`
+}
+
+func detectWpTheme(c *base) (vinfos []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp theme list --path=%s --format=json", c.ServerInfo.WpPath)
+
+	var themes []WpStatus
+	var r execResult
+	if r = exec(c.ServerInfo, cmd, noSudo); r.isSuccess() {
+		if err = json.Unmarshal([]byte(r.Stdout), &themes); err != nil {
+			return
+		}
+	}
+	if !r.isSuccess() {
+		return vinfos, fmt.Errorf("%s", cmd)
+	}
+
+	for _, theme := range themes {
+		url := fmt.Sprintf("https://wpvulndb.com/api/v3/themes/%s", theme.Name)
+		token := fmt.Sprintf("Token token=%s", c.ServerInfo.WpToken)
+		var req *http.Request
+		req, err = http.NewRequest("GET", url, nil)
+		if err != nil {
+			return
+		}
+		req.Header.Set("Authorization", token)
+		client := new(http.Client)
+		var resp *http.Response
+		resp, err = client.Do(req)
+		if err != nil {
+			return
+		}
+		body, _ := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 && resp.StatusCode != 404 {
+			return vinfos, fmt.Errorf("status: %s", resp.Status)
+		} else if resp.StatusCode == 404 {
+			var jsonError WpCveInfos
+			if err = json.Unmarshal([]byte(string(body)), &jsonError); err != nil {
+				return
+			}
+			if jsonError.Error == "HTTP Token: Access denied.\n" {
+				return vinfos, fmt.Errorf("wordpress: HTTP Token: Access denied.")
+			} else if jsonError.Error == "Not found" {
+				c.log.Infof("wordpress: %s not found", theme.Name)
+			} else {
+				return vinfos, fmt.Errorf("status: %s", resp.Status)
+			}
+		}
+		contentConvertVinfos(string(body), theme)
+	}
+	return
+}
+
+func detectWpPlugin(c *base) (vinfos []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp plugin list --path=%s --format=json", c.ServerInfo.WpPath)
+
+	var plugins []WpStatus
+	var r execResult
+	if r := exec(c.ServerInfo, cmd, noSudo); r.isSuccess() {
+		if err = json.Unmarshal([]byte(r.Stdout), &plugins); err != nil {
+			return
+		}
+	}
+	if !r.isSuccess() {
+		return vinfos, fmt.Errorf("%s", cmd)
+	}
+
+	for _, plugin := range plugins {
+		url := fmt.Sprintf("https://wpvulndb.com/api/v3/plugins/%s", plugin.Name)
+		token := fmt.Sprintf("Token token=%s", c.ServerInfo.WpToken)
+		var req *http.Request
+		req, err = http.NewRequest("GET", url, nil)
+		if err != nil {
+			return
+		}
+		req.Header.Set("Authorization", token)
+		client := new(http.Client)
+		var resp *http.Response
+		resp, err = client.Do(req)
+		if err != nil {
+			return
+		}
+		body, _ := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 && resp.StatusCode != 404 {
+			return vinfos, fmt.Errorf("status: %s", resp.Status)
+		} else if resp.StatusCode == 404 {
+			var jsonError WpCveInfos
+			if err = json.Unmarshal([]byte(string(body)), &jsonError); err != nil {
+				return
+			}
+			if jsonError.Error == "HTTP Token: Access denied.\n" {
+				return vinfos, fmt.Errorf("wordpress: HTTP Token: Access denied.")
```

Comment:
__[Golint]__ error strings should not be capitalized or end with punctuation or a newline

<sub>posted by Sider</sub>